gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/SubscriptionServiceImpl.java

```diff
@@ -1608,14 +1599,30 @@ public String exportAsCsv(Collection<SubscriptionEntity> subscriptions, Map<Stri
 
             if (subscription.getEndingAt() != null) {
                 sb.append(dateFormatter.format(subscription.getEndingAt()));
-                sb.append(separator);
             }
+            sb.append(separator);
 
             sb.append(subscription.getStatus());
 
             sb.append(lineSeparator());
```

Comment:
I think that we can have more maintainable code with something like:

```java
            Collection<String> cells = new ArrayList<>();
            cells.add(getName(plan));
            final Object application = metadata.get(subscription.getApplication());
            cells.add(getName(application));
            if (subscription.getCreatedAt() != null) {
                cells.add(dateFormatter.format(subscription.getCreatedAt()));
            }
            if (subscription.getProcessedAt() != null) {
                cells.add(dateFormatter.format(subscription.getProcessedAt()));
            }
            if (subscription.getStartingAt() != null) {
                cells.add(dateFormatter.format(subscription.getStartingAt()));
            }

            if (subscription.getEndingAt() != null) {
                cells.add(dateFormatter.format(subscription.getEndingAt()));
            }

            cells.add(subscription.getStatus());

            sb.append(String.joining(separator, cells).append(lineSeparator());
```

Reply:
Agreed, we can make it more maintainable. There is no `String.joining(...)` method, maybe you got something else on you mind?

Reply:
Hum... I'm surprising I think about it https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/String.html#join(java.lang.CharSequence,java.lang.Iterable)

🤔

Reply:
Yes I thought you wanted to use `join` but wanted to be sure. `join` accepts `CharSequence` or `Iterable<? extends CharSequence>` but doesn't accept `List<String>`. Do you know any elegant solution to apply `cells` to `join` method?

Reply:
`List<String>` (or `Collection<String>`) is a Ìterable<CharSequence>` you have some example in apim like here

https://github.com/gravitee-io/gravitee-api-management/blob/master/gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/ApplicationAlertServiceImpl.java#L130

Reply:
Ok errors were about `separator` not the `cells` itself... Now it's correct.

```java
    private static final String separator = ";";
    ...
    sb.append(String.join(String.join(separator, cells))).append(lineSeparator());

```

Reply:
Ok, unfortunately proposed solution does now work at all because we have to add separator for every null value to keep CSV columns structure. I would refactor this in the following way.

```java
private void prepareSubscriptionCsvRows(
        Collection<SubscriptionEntity> subscriptions,
        Map<String, Map<String, Object>> metadata,
        StringBuilder sb
    ) {
        for (final SubscriptionEntity subscription : subscriptions) {
            final Object plan = metadata.get(subscription.getPlan());
            Collection<String> cells = new ArrayList<>();
            cells.add(getName(plan));
            final Object application = metadata.get(subscription.getApplication());
            cells.add(getName(application));

            prepareDateCell(subscription.getCreatedAt(), cells);
            prepareDateCell(subscription.getProcessedAt(), cells);
            prepareDateCell(subscription.getStartingAt(), cells);
            prepareDateCell(subscription.getEndingAt(), cells);

            cells.add(subscription.getStatus().name());
            sb.append(String.join(separator, cells)).append(lineSeparator());
        }
    }

    private void prepareDateCell(Date date, Collection<String> cells) {
        if (date != null) {
            cells.add(dateFormatter.format(date));
        } else {
            cells.add("");
        }
    }
```
What do you think?